lib/js/jsffi.nim

```diff
@@ -0,0 +1,479 @@
+
+proc `.=`*[V](obj: JsAssoc[string, V], field: cstring, value: V) =
+  ## Experimental dot accessor (set) for type JsAssoc.
+  ## Sets the value of a property of name `field` in a JsObject `x` to `value`.
+  obj.setImpl(field, value)
+
+macro `.()`*[V: proc](obj: JsAssoc[string, V], field: cstring, args: varargs[untyped]): auto =
+  ## Experimental "method call" operator for type JsAssoc.
+  ## Takes the name of a method of the JavaScript object (`field`) and calls
+  ## it with `args` as arguments. Here, everything is typechecked, so you do not
+  ## have to worry about `undefined` return values.
+  result = quote do:
+    let temp = getAsImpl[cstring, `obj`.V](`obj`, `field`)
+    temp()
+  for elem in args:
+    result[1].add elem
+
+# Iterators:
+
+iterator pairs*(obj: JsObject): (cstring, JsObject) =
+  ## Yields tuples of type ``(cstring, JsObject)``, with the first entry
+  ## being the `name` of a fields in the JsObject and the second being its
+  ## value wrapped into a JsObject.
+  var k: cstring
+  var v: JsObject
+  {.emit: "for (var `k` in `obj`) {".}
+  {.emit: "  if (!`obj`.hasOwnProperty(`k`)) continue;".}
+  {.emit: "  `v`=`obj`[`k`];".}
+  yield (k, v)
+  {.emit: "}".}
+
+iterator items*(obj: JsObject): JsObject =
+  ## Yields the `values` of each field in a JsObject, wrapped into a JsObject.
+  var v: JsObject
+  {.emit: "for (var k in `obj`) {".}
+  {.emit: "  if (!`obj`.hasOwnProperty(k)) continue;".}
+  {.emit: "  `v`=`obj`[k];".}
+  yield v
+  {.emit: "}".}
+
+iterator keys*(obj: JsObject): cstring =
+  ## Yields the `names` of each field in a JsObject.
+  var k: cstring
+  {.emit: "for (var `k` in `obj`) {".}
+  {.emit: "  if (!`obj`.hasOwnProperty(`k`)) continue;".}
+  yield k
+  {.emit: "}".}
+
+iterator pairs*[K, V](assoc: JsAssoc[K, V]): (K,V) =
+  ## Yields tuples of type ``(K, V)``, with the first entry
+  ## being a `key` in the JsAssoc and the second being its corresponding value.
+  when K is string:
+    var k: cstring
+  else:
+    var k: K
+  var v: V
+  {.emit: "for (var `k` in `assoc`) {".}
+  {.emit: "  if (!`assoc`.hasOwnProperty(`k`)) continue;".}
+  {.emit: "  `v`=`assoc`[`k`];".}
+  when K is string:
+    yield ($k, v)
+  else:
+    yield (k, v)
+  {.emit: "}".}
+
+iterator items*[K,V](assoc: JSAssoc[K,V]): V =
+  ## Yields the `values` in a JsAssoc.
+  var v: V
+  {.emit: "for (var k in `assoc`) {".}
+  {.emit: "  if (!`assoc`.hasOwnProperty(k)) continue;".}
+  {.emit: "  `v`=`assoc`[k];".}
+  yield v
+  {.emit: "}".}
+
+iterator keys*[K,V](assoc: JSAssoc[K,V]): K =
+  ## Yields the `keys` in a JsAssoc.
+  when K is string:
+    var k: cstring
+  else:
+    var k: K
+  {.emit: "for (var `k` in `assoc`) {".}
+  {.emit: "  if (!`assoc`.hasOwnProperty(`k`)) continue;".}
+  when K is string:
+    yield $k
+  else:
+    yield k
+  {.emit: "}".}
+
+# Literal generation
+
+macro lit*(xs: varargs[untyped]): auto =
+  ## Takes a ``typedesc`` as its first argument, and a series of expressions of
```

Comment:
Now takes a `typedesc` as its first argument and an `untyped` as its second.
The rationale behind this is to stay as close as possible to the previous syntax, which is not achievable using `typedesc` and `varargs[untyped]`, as an expression `Typ.lit(a = 11, b = 22)` is parsed as a call with named arguments, and not a `varargs[untyped]`, which is why a macro with those argument types could not be called.